File: src/core/Economy.js
/**
 * Economy — custos, produção, buy modes e milestones (spec §27–§30, §76–§77).
 * Engine-agnóstico. API pública espelhada em `unity/Economy.cs`.
 */
import BigNumber from './BigNumber.js';

export const BUY_MODES = [1, 10, 25, 100, -1]; // -1 = MAX

/** Custo de comprar `qty` unidades do produtor, já possuindo `owned`. */
export function buyCost(baseCost, growthRate, owned, qty) {
  const base = BigNumber.fromNumber(baseCost);
  const g = BigNumber.fromNumber(growthRate);
  // soma geométrica fechada: baseCost * r^owned * (r^qty - 1)/(r - 1)
  const rPowOwned = g.pow(owned);
  const rPowQty = g.pow(qty);
  const numerator = rPowQty.sub(BigNumber.one());
  const denominator = BigNumber.fromNumber(growthRate - 1);
  return base.mul(rPowOwned).mul(numerator).div(denominator);
}

/** Maior quantidade comprável com `balance` (MAX) — fechado com progressão geométrica. */
export function maxAffordable(baseCost, growthRate, owned, balance) {
  const base = BigNumber.fromNumber(baseCost);
  const g = BigNumber.fromNumber(growthRate);
  const denom = BigNumber.fromNumber(growthRate - 1);
  // balance * (r-1) / (base * r^owned) + 1 >= r^n  =>  n <= log_r( ... )
  const budget = balance.mul(denom).div(base.mul(g.pow(owned))).add(BigNumber.one());
  if (budget.lte(BigNumber.one())) return 0;
  const n = Math.floor(budget.log10() / Math.log10(growthRate));
  // ajuste fino por iteração limitada (robustez numérica)
  let best = 0;
  for (let cand = Math.max(0, n - 1); cand <= n + 1; cand++) {
    const cost = buyCost(baseCost, growthRate, owned, cand);
    if (cost.lte(balance)) best = cand;
    else break;
  }
  return best;
}

/** Multiplicador acumulado de milestones para `owned` unidades (§30). */
export function milestoneMultiplier(milestones, owned) {
  let mult = BigNumber.one();
  for (const m of milestones) {
    if (owned >= m.count) mult = mult.scale(m.factor);
  }
  return mult;
}

/** Produção por segundo de um produtor. */
export function producerProduction(baseProduction, owned, milestones, ownedCount) {
  const base = BigNumber.fromNumber(baseProduction);
  const mmult = milestoneMultiplier(milestones, ownedCount);
  return base.scale(owned).mul(mmult);
}

export default { BUY_MODES, buyCost, maxAffordable, milestoneMultiplier, producerProduction };

File: unity/Economy.cs
// Economy — espelho C# (Unity-ready) de src/core/Economy.js
// Custos, produção, buy modes e milestones (spec §27–§30). Porta 1:1.
using System;

namespace ODespertou.Core
{
    public static class Economy
    {
        public static readonly int[] BuyModes = { 1, 10, 25, 100, -1 }; // -1 = MAX

        public static BigNumber BuyCost(double baseCost, double growthRate, long owned, long qty)
        {
            var baseBn = BigNumber.FromNumber(baseCost);
            var g = BigNumber.FromNumber(growthRate);
            var num = g.Pow(qty).Sub(BigNumber.One);
            var den = BigNumber.FromNumber(growthRate - 1);
            return baseBn.Mul(g.Pow(owned)).Mul(num).Div(den);
        }

        public static long MaxAffordable(double baseCost, double growthRate, long owned, BigNumber balance)
        {
            var baseBn = BigNumber.FromNumber(baseCost);
            var g = BigNumber.FromNumber(growthRate);
            var den = BigNumber.FromNumber(growthRate - 1);
            var budget = balance.Mul(den).Div(baseBn.Mul(g.Pow(owned))).Add(BigNumber.One);
            if (budget.Lte(BigNumber.One)) return 0;
            long n = (long)Math.Floor(budget.Log10() / Math.Log10(growthRate));
            long best = 0;
            for (long cand = Math.Max(0, n - 1); cand <= n + 1; cand++)
            {
                if (BuyCost(baseCost, growthRate, owned, cand).Lte(balance)) best = cand;
                else break;
            }
            return best;
        }
    }

    public class MilestoneDef
    {
        public long Count;
        public double Factor;
    }

    public static class Milestones
    {
        public static BigNumber Multiplier(MilestoneDef[] defs, long owned)
        {
            var mult = BigNumber.One;
            foreach (var d in defs) if (owned >= d.Count) mult = mult.Scale(d.Factor);
            return mult;
        }

        public static BigNumber Production(double baseProduction, long owned, MilestoneDef[] defs)
        {
            return BigNumber.FromNumber(baseProduction).Scale(owned).Mul(Multiplier(defs, owned));
        }
    }
}
